keystore/src/connection/mod.rs

```diff
@@ -227,10 +277,21 @@ impl FetchFromDatabase for Connection {
     ) -> CryptoKeystoreResult<Vec<E>> {
         let entity_ids: Vec<StringEntityId> = ids.iter().map(|id| id.as_slice().into()).collect();
         let mut conn = self.conn.lock().await;
-        E::find_many(&mut conn, &entity_ids).await
+        let persisted_records = E::find_many(&mut conn, &entity_ids).await?;
+
+        let transaction_guard = self.transaction.lock().await;
+        let Some(transaction) = transaction_guard.as_ref() else {
+            return Ok(persisted_records);
+        };
+        transaction.find_many(persisted_records, ids).await
     }
 
     async fn count<E: Entity<ConnectionType = KeystoreDatabaseConnection>>(&self) -> CryptoKeystoreResult<usize> {
+        if self.transaction.lock().await.is_some() {
+            // Unfortunately, we have to do this because of possible record id overlap 
+            // between cache and db.
+            return Ok(self.find_all::<E>(Default::default()).await?.len());
+        };
```

Comment:
This looks suspicious. Isn't `find_all` going to try to acquire the same lock that we got here?

Reply:
The tests run and execute this codepath without a deadlock. Calling `.is_some()` probably drops the lock.